modules/mongodb/collect.go

```diff
@@ -1,181 +1,49 @@
 package mongo
 
-import (
-	"fmt"
-	"reflect"
+import "fmt"
 
-	"github.com/netdata/go.d.plugin/agent/module"
-	"github.com/netdata/go.d.plugin/pkg/stm"
-)
-
-// collectServerStatus creates the map[string]int64 for the available dims.
-// nil values will be ignored and not added to the map and thus metrics
-// should not appear on the dashboard.
-// Because mongo reports a metric only after it first appears, some dims might
-// take a while to appear. For example, in order to report number of create
-// commands, a document must be created first.
-func (m *Mongo) collectServerStatus(ms map[string]int64) error {
-	status, err := m.mongoCollector.serverStatus()
-	if err != nil {
-		return fmt.Errorf("serverStatus command failed: %s", err)
+func (m *Mongo) collect() (map[string]int64, error) {
+	if err := m.mongoCollector.initClient(m.URI, m.Timeout); err != nil {
+		return nil, fmt.Errorf("init mongo client: %v", err)
 	}
 
-	m.addOptionalCharts(status)
-	for k, v := range stm.ToMap(status) {
-		ms[k] = v
+	ms := map[string]int64{}
+	if err := m.collectServerStatus(ms); err != nil {
+		return nil, fmt.Errorf("couldn't collecting server status metrics: %s", err)
 	}
-	return nil
-}
 
-// addOptionalCharts tries to add charts based on the availability of
-// metrics coming back from the `serverStatus` command.
-// Nil pointer structs will be skipped and we won't produce metrics for
-// unavailable metrics.
-func (m *Mongo) addOptionalCharts(status *serverStatus) {
-	m.metricExists(status.Transactions, &chartTransactionsCurrent)
-	m.metricExists(status.FlowControl, &chartFlowControl)
-
-	if status.GlobalLock != nil {
-		m.metricExists(status.GlobalLock.ActiveClients, &chartGlobalLockActiveClients)
-		m.metricExists(status.GlobalLock.CurrentQueue, &chartGlobalLockCurrentQueue)
-	}
-	if status.Tcmalloc != nil {
-		m.metricExists(status.Tcmalloc.Generic, &chartTcmallocGeneric)
-		m.metricExists(status.Tcmalloc.Tcmalloc, &chartTcmalloc)
+	if err := m.collectDbStats(ms); err != nil {
+		m.Errorf("couldn't collecting dbstats metrics: %s", err)
```

Comment:
We don't return "nil, err" here, is that on purpose?

Reply:
my idea is to return at least the serverStatus metric if they have been collected successfully before this statement. I am not sure if it's correct or not. Do  you think we should return (nil, error) in such cases?

Reply:
> my idea is to return at least the serverStatus 

Yeah, that makes sense. Then we need to return "ms, error" and remove https://github.com/georgeok/go.d.plugin/blob/2c3c75bd5dde92d9c28074f7d9b271e92f7ca59e/modules/mongodb/mongodb.go#L94

Reply:
done